### ISSUE TITLE
Improve test stability of t.w.test.ContentDecoderAgentTests.test_existingHeaders

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7680]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)| @multani reported|
|-|-|
|Trac ID|trac#7680|
|Type|defect|
|Created|2014-10-14 09:27:56Z|
|Branch|https://github.com/twisted/twisted/tree/contentdecoderagent-test-stability-7680|

The test at https://twistedmatrix.com/trac/browser/trunk/twisted/web/test/test_agent.py?rev=43358#L1941 uses `Headers.getAllRawHeaders()` which result is case as a list and compared with another list with more than 1 element.

On my machine, this test (sometimes) fails with the following:

```
twisted.test.test_sslverify.OpenSSLOptions.test_inspectCertificate
===============================================================================
[FAIL]
Traceback (most recent call last):
  File ".../twisted/twisted/web/test/test_agent.py", line 1961, in test_existingHeaders
    ('Accept-Encoding', ['fizz', 'decoder1,decoder2'])])
  File ".../twisted/twisted/trial/_synctest.py", line 447, in assertEqual
    % (msg, pformat(first), pformat(second)))
twisted.trial.unittest.FailTest: not equal:
a = [('Host', ['example.com']),
 ('Accept-Encoding', ['fizz', 'decoder1,decoder2']),
 ('Foo', ['bar'])]
b = [('Host', ['example.com']),
 ('Foo', ['bar']),
 ('Accept-Encoding', ['fizz', 'decoder1,decoder2'])]
```

The problem comes from the fact that `Headers.getAllRawHeaders()` doesn't guarantee the order of the data it returns (since internally it iterates over a dictionary) and for some reason, here the expected result doesn't have the same ordering as the actual result.

The `getAllRawHeaders` API doesn't say anything about the ordering, so I'm assuming it doesn't really care about it, and I'm proposing a simple fix which sorts the result returned before comparing it in the test.

Attachments:

* [7680-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6ee/6eea5216ce169fd6753767b1130eefdf45e5141f/debdfd633174901517f42df4c701c53f829660a0.patch) (912 bytes) - added by multani on 2014-10-14 09:31:27Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7680 7680
type__defect defect
reporter__multani multani
priority__normal normal
milestone__ 
branch__branches_contentdecoderagent_test_stability_7680 branches/contentdecoderagent-test-stability-7680
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1413278876344057 1413278876344057
changetime__1413282226939022 1413282226939022
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_1"></a>|@multani commented|
|-|-|

Trac doesn't show it, but I put a .misc topfile in `twisted/web/topfile/7680.misc` in the patch.

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_2"></a>|@multani commented|
|-|-|

(for quicker reference, here the implementation of `getAllRawHeaders()` : http://twistedmatrix.com/trac/browser/tags/releases/twisted-14.0.2/twisted/web/http_headers.py#L237 )

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl set owner to @hawkowl|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

(In [43359]) Branching to contentdecoderagent-test-stability-7680.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

(In [43360]) applying patch from multani, refs [#7680](https://github.com/twisted/twisted/issues/7680)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

Some builders are being cranky, but they're unrelated. LGTM.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl set status to `closed`|
|-|-|

(In [43362]) Merging contentdecoderagent-test-stability-7680: Improve stability of t.w.test.ContentDecoderAgentTests.test_existingHeaders

Author: multani
Reviewer: hawkowl
Fixes: [#7680](https://github.com/twisted/twisted/issues/7680)
